Rdrp 1098 unit tests imputation helpers

ewingr42 opened the pull request:
## Pull Request submission

* The first new imputation_helper test is included: get_imputation_cols

* pull request is for info as there are 10 more tests to complete 

* I have included a set of test data that is different to the live system

#### Closes or fixes

* ticket RDRP-1098
initial commit

#### Code

- [ y] **Code runs** The code runs on my machine
- [ y] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ partial] **Requirements** My/our code functions according to the requirements of the ticket
- [ na] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ na] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ y] **Clean Code**
    - [ y] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [y ] Code passess flake8 linting check
    - [ y] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ n, fixed in next release] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)


#### Documentation

Any new code includes all the following forms of documentation:

- [y ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [na ] Includes `Args` and `returns` for all major functions
    - [na] The docstring details data types
- [ na] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [y] All data needed to run this script is available in Dev/Test
- [na] All data is excluded from this pull request
- [ y] Secrets checker pre-commit passes

#### Testing
- [na] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_


### Review comments

AnneONS committed:
Merge remote-tracking branch 'origin/develop' into RDRP-947_PNP_QA_output

ewingr42 committed:
Merge remote-tracking branch 'origin' into RDRP-1098_unit_tests_imputation_helpers

ewingr42 committed:
test imputation_helpers get_imputation_cols update

Ryan2Y79 commented:
Is it advised by Keilan Evans that the preposed changes to `.pre-commit-config.yaml` be committed to develop so that these changes propagate to all devs and users of the pipeline.

AnneONS commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -412,3 +413,46 @@ def test_imputation_marker(self):
+                                         "headcount_tot_y", "headcount_total"}
+                            }
+                  }
+        return input_config
```
This unit test works well.

The test files don't go through the pre-commit, which makes us freer, for example we don't need to keep to the 88 character rule. However, the style that we use in the pipe line that the precommit would normally enforce is that every  open bracket follows a new line. You've partially done this here, but our style would be to do this (which makes it really easy to see if you've missed off closing brackets-  a coder's nightmare as I'm sure you know)

       input_config = {
            "breakdowns" : {
                "219" : {
                    "202", "203", "204", "205", "206", "207", "209",
                    "210", "212", "214", "216", "218", "219", "220",
                    "221", "222", "223", "225", "226", "227", "228",
                    "229", "237", "242", "243", "244", "245", "246",
                    "247", "248", "249", "250", "305", "302", "399"
                },
                "emp_total" : {
                    "emp_researcher", "emp_technician", "emp_other"
                },
                "headcount_total" : {
                    "headcount_res_m", "headcount_res_f",
                    "headcount_tec_m", "headcount_tec_f",
                    "headcount_oth_x", "headcount_oth_y"
                }
            },
            "imputation" : {
                "sum_cols": {
                    "emp_total", "headcount_tot_x",
                    "headcount_tot_y", "headcount_total"
                }
            }
        }

AnneONS commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -412,3 +413,46 @@ def test_imputation_marker(self):
+                           "emp_researcher", "emp_technician", "emp_other",
+                           "headcount_res_m", "headcount_res_f", "headcount_tec_m",
+                           "headcount_tec_f", "headcount_oth_x", "headcount_oth_y",
+                           "headcount_tot_x", "headcount_tot_y"]
```
I like the way you've set out the test with the expected output.

Just a suggestion, it might be easier to check that the columns match the config if they were presented in the same order, or something.

ewingr42 replied:
thanks for the review and the advice, I'll update the code when I do the next test.

AnneONS reviewed (commented):
Great! Look forward to seeing the other tests.

ewingr42 committed:
test imputation_helpers create_imp_class_col

ewingr42 commented:
I've added the test for create_imp_class_col (I believe)...

AnneONS commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -412,3 +413,122 @@ def test_imputation_marker(self):
+        run2 = create_imp_class_col(run1, column_list=['201', 'rusic'], class_name='pg_class', use_cellno=True)
+        result_df = create_imp_class_col(run2, column_list=['201'], class_name='pg_sic_class', use_cellno=True)
+
+        assert_frame_equal(result_df.reset_index(drop=True), exp_output_df)
```
HI Rob, thanks for this new test.
We already have tests for this function, don't know if you saw those. However, you have usefully added new test cases for when we use the "pg_class" and "pg_sic_class".

If you'd like to chat some time, I'll talk you through how I would have approached this task.

AnneONS reviewed (commented):
Thanks for this next test- it looks good, and I'm happy to discuss

ewingr42 committed:
replace test helpers file with develop version

ewingr42 committed:
Merge branch 'RDRP-1098_unit_tests_imputation_helpers' of https://github.com/ONSdigital/research-and-development into RDRP-1098_unit_tests_imputation_helpers

ewingr42 committed:
check r and d inst

ewingr42 committed:
check r and d inst simplify references

AnneONS commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -450,3 +453,168 @@ def test_imputation_marker(self):
+        values_df3 = [
+            [True, 5],
+            [False, 6]
+        ]
```
A style thing: we use _df to indicate a dataframe, rather than lists (of lists) - I can see your variable name means "columns for df1" but columns1 would be more intuitive if you're used to thinking " _df means it's a dataframe"

AnneONS commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -450,3 +453,168 @@ def test_imputation_marker(self):
+        df1, df2, df3 = self.input_dfs()
+        expected_df = self.expected_output()
+
+        result_df = concat_with_bool([df1, df2, df3])
```
this test is useful, thanks. Could we also check if the function works as expected if we have nulls in the input df cols?

AnneONS commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -450,3 +453,168 @@ def test_imputation_marker(self):
+
+        # Create DataFrame from the lists of values
+        df = pd.DataFrame(values_df, columns=columns_df)
+        return df
```
again, style: better to reserve _df for when it's a dataframe

AnneONS reviewed (commented):
just a partial review for one of the tests so far

ewingr42 committed:
get_imputation_cols test and tidy up

ewingr42 committed:
imputation_helpers test instance_fix

ewingr42 committed:
updated concat_with_bool test

woodac commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -450,3 +396,286 @@ def test_imputation_marker(self):
+        check_final_df, check_mult_604_qa_df = create_r_and_d_instance(input_df)
+
+        assert_frame_equal(check_final_df.reset_index(drop=True), expected_final_df, check_dtype=False)
+        assert_frame_equal(check_mult_604_qa_df.reset_index(drop=True), expected_mult_604_qa_df, check_dtype=False)
```
My interpretation of this function is that it is addressing references that only have 1 entry, that fit the criteria of long forms, no R&D  with an instance 0. Maybe you could change a couple of the values in the input data to reflect this so it it easier to see if the function is creating a duplicate row and setting the instance to 1 as the expected outcome.

AnneONS replied:
this test is great, thanks.

woodac commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -450,3 +396,286 @@ def test_imputation_marker(self):
+        expected_output = self.cre_expected_output()
+
+        output = create_notnull_mask(input_df, "601")
+        assert_series_equal(output, expected_output)
```
Looks good :)

ewingr42 committed:
updated test data for create_r_and_d_instance test_create_mask and updated helper function create mask

ewingr42 commented:
I created a test for create_mask... I put some fixes into the helper function but the test should work with the fixed function if this has been done elsewhere. I added tests for the single options, and a couple of examples of multiple options. It would be simple to add extra test data, if required, as they are simple 'or's of the single options.
I also fixed the create_r_and_d_instance function. This may be a duplicate, if so, sorry. But it was helpful to me in any case.

AnneONS commented on src/imputation/imputation_helpers.py:
```diff
@@ -99,22 +102,22 @@ def create_mask(df: pd.DataFrame, options: List) -> pd.Series:
 
     if "clear_status" in options:
-        df["mask_col"] = df["mask_col"] & clear_mask
+        df["mask_col"] = df["mask_col"] | clear_mask
```
Hi Rob, the reason I put & here rather than | is that I want the option to filter on several conditions at the same time. 
So rather than having "it is a clear status or it is long form", I want "it is clear status and it is long form". 

This would be written 

(df["formtype"] == "0001")  & (df["status"].isin(["Clear", "Clear - overridden"]))

AnneONS commented on src/imputation/imputation_helpers.py:
```diff
@@ -99,22 +102,22 @@ def create_mask(df: pd.DataFrame, options: List) -> pd.Series:
     df["mask_col"] = False
```
This code has changed quite a lot recently- can you merge in develop?

AnneONS commented on src/imputation/imputation_helpers.py:
```diff
@@ -351,6 +354,37 @@ def split_df_on_imp_class(df: pd.DataFrame, exclusion_list: List = ["817", "nan"
+        & (df["instance"] != 0)
+        & (df["status"].isin(["Clear", "Clear - overridden"]))
+        & excl_postcode_only_mask
+    )
```
do you think it's worth using the create_mask function here?

ewingr42 replied:
i have removed my version of the create_mask test

AnneONS commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -450,3 +397,341 @@ def test_imputation_marker(self):
+            [1006, 2, "Yes", "0001"],
+            [1003, 0, "No", "0001"],
+            [1003, 1, "No", "0001"],
+        ]
```
Perhaps I'm lazy, but I'd find it easier to focus on what's going on here if the data was organised by "reference" then "instance". Thanks!

ewingr42 replied:
the function includes a 'sort' so I created the test data 'unsorted'. But I can easily tidy it up :)

AnneONS replied:
good point about the sort. It's much easier to read now, though, and follow what the func is doing, Thanks!

AnneONS replied:
The sort is needed as we create a new dataframe for the extra rows we are adding and then concatenate. But the original input dataframe would be already sorted as we now have it.

AnneONS reviewed (commented):
a couple more comments- could you merge in develop?
do a commit or a git stash so you don't have any pending changes, then
git fetch
git merge origin develop
Give one of us a shout and we'll look at the merge conflicts together

ewingr42 committed:
Merge remote-tracking branch origin into RDRP-1098_unit_tests_imputation_helpers

ewingr42 committed:
tidied up input data for create_r_and_d_instance

woodac commented on src/imputation/imputation_helpers.py:
```diff
@@ -37,7 +37,7 @@ def get_imputation_cols(config: dict) -> list:
 def create_imp_class_col(
     df: pd.DataFrame,
-    column_list: List[str],
+    column_list: list[str],
```
Why is this changing when the script imports 'List'?

AnneONS replied:
@ewingr42 James got us using special types from the Typing import- List is different from the usual python list and I think allows you do do more with arguments.

ewingr42 committed:
remove instance fix test

AnneONS commented on tests/test_imputation/test_imputation_helpers.py:
```diff
@@ -489,10 +436,13 @@ def expected_output(self):
+            [False, False, False, False, True, 7],
+            [False, False, False, False, False, 8],
+            [False, False, False, False, False, 9],
         ]
```
This is a nicely designed test.

Just thought of one more thing we might want to test, the case where one of the bool columns occurs in more than one dataframe

AnneONS reviewed (commented):
Nearly there now, just a couple of minor things.

ewingr42 committed:
included 211_trim in df1 as example of bool column in more than one df

AnneONS committed:
Merge remote-tracking branch 'origin' into RDRP-1098_unit_tests_imputation_helpers

AnneONS reviewed (approved):
Fab! Let's get this merged.
